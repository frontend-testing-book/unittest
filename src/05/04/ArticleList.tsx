import type { ItemProps } from './ArticleListItem'
import { ArticleListItem } from './ArticleListItem'

type Props = {
  items: ItemProps[]
}

export const ArticleList = ({ items }: Props) => {
  return (
    <div>
      <h2>記事一覧</h2>
      {items.length ? (
        <ul>
          {items.map((item) => (
            <ArticleListItem {...item} key={item.id} />
          ))}
        </ul>
      ) : (
        <p>投稿記事がありません</p>
      )}
    </div>
  )
}
