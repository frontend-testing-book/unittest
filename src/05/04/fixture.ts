import type { ItemProps } from './ArticleListItem'

export const items: ItemProps[] = [
  {
    id: 'howto-testing-with-typescript',
    title: 'TypeScript を使ったテストの書き方',
    body: 'テストを書く時、TypeScript を使うことで、テストの保守性が向上します…',
  },
  {
    id: 'nextjs-link-component',
    title: 'Next.js の Link コンポーネント',
    body: 'Next.js の画面遷移には、Link コンポーネントを使用します…',
  },
  {
    id: 'react-component-testing-with-jest',
    title: 'Jest ではじめる React のコンポーネントテスト',
    body: 'Jest は単体テストとして、UIコンポーネントのテストが可能です…',
  },
]
