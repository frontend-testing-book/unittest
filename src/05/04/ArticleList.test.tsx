import { render, screen, within } from '@testing-library/react'

import { ArticleList } from './ArticleList'
import { items } from './fixture'

test('タイトルの表示', () => {
  render(<ArticleList items={items} />)
  expect(screen.getByRole('heading', { name: '記事一覧' })).toBeInTheDocument()
})

test('items の数だけ一覧表示される', () => {
  render(<ArticleList items={items} />)
  expect(screen.getAllByRole('listitem')).toHaveLength(3)
})

test('items の数だけ一覧表示される', () => {
  render(<ArticleList items={items} />)
  const list = screen.getByRole('list')
  expect(list).toBeInTheDocument()
  expect(within(list).getAllByRole('listitem')).toHaveLength(3)
})

test('一覧アイテムが空のとき「投稿記事がありません」が表示される', () => {
  render(<ArticleList items={[]} />)
  const list = screen.queryByRole('list')
  expect(list).not.toBeInTheDocument()
  expect(list).toBeNull()
  expect(screen.getByText('投稿記事がありません')).toBeInTheDocument()
})

test('Snapshot: items の数だけ一覧表示される', () => {
  const { container } = render(<ArticleList items={items} />)
  expect(container).toMatchSnapshot()
})
