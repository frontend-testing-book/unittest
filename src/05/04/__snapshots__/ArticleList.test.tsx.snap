// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: items の数だけ一覧表示される 1`] = `
<div>
  <div>
    <h2>
      記事一覧
    </h2>
    <ul>
      <li>
        <h3>
          TypeScript を使ったテストの書き方
        </h3>
        <p>
          テストを書く時、TypeScript を使うことで、テストの保守性が向上します…
        </p>
        <a
          style="color: blue; cursor: pointer; text-decoration: underline;"
        >
          もっと見る
        </a>
      </li>
      <li>
        <h3>
          Next.js の Link コンポーネント
        </h3>
        <p>
          Next.js の画面遷移には、Link コンポーネントを使用します…
        </p>
        <a
          style="color: blue; cursor: pointer; text-decoration: underline;"
        >
          もっと見る
        </a>
      </li>
      <li>
        <h3>
          Jest ではじめる React のコンポーネントテスト
        </h3>
        <p>
          Jest は単体テストとして、UIコンポーネントのテストが可能です…
        </p>
        <a
          style="color: blue; cursor: pointer; text-decoration: underline;"
        >
          もっと見る
        </a>
      </li>
    </ul>
  </div>
</div>
`;
